yahp/utils/json_schema_helpers.py

```diff
@@ -49,43 +73,84 @@ def get_type_json_schema(f_type: type_helpers.HparamsType):
     elif f_type.type == type_helpers._JSONDict:
         res = {
             'type': 'object',
         }
     # Hparam class
     elif callable(f_type.type):
-        hparam_class = ensure_hparams_cls(f_type.type)
-        res = hparam_class.get_json_schema()
+        # If the parent class was autoyahped, do not try autoyahping parameters
+        if from_autoyahp:
+            res = {
+                'type': 'object',
+            }
+        # Otherwise, attempt to autoyahp
+        else:
+            hparam_class = ensure_hparams_cls(f_type.type)
+            # Build schema and add to _cls_def if not present. get_json_schema adds to _cls_def
+            # internally, so we only need to call the function.
+            if hparam_class not in _cls_def:
+                hparam_class.get_json_schema(_cls_def)
+            _cls_def[hparam_class.__name__]['referenced'] = True
+            res = {'$ref': f'#/$defs/{hparam_class.__name__}'}
     else:
         raise ValueError('Unexpected type when constructing JSON Schema.')
 
-    return _check_for_list_and_optional(f_type, res)
+    return _check_for_list_and_optional(f_type, res, _cls_def)
 
 
-def _check_for_list_and_optional(f_type: type_helpers.HparamsType, schema: Dict[str, Any]) -> Dict[str, Any]:
+def _check_for_list_and_optional(f_type: type_helpers.HparamsType, schema: Dict[str, Any],
+                                 _cls_def: Dict[str, Any]) -> Dict[str, Any]:
     """Wrap JSON Schema with list schema or optional schema if specified.
     """
-    res = schema
+    if not f_type.is_list and not f_type.is_optional:
+        return schema
+
+    # Use defs to avoid duplicate schema for more complex types
+    if f_type.is_list and (len(f_type.types) > 1 or f_type.type not in (str, bool, int, float)):
+        # Use counter to give unique key name. While enums and hparams can generate unique names
+        # based on classes, lists may be of primitive types, meaning we can't generate unique
+        # names.
+        _cls_def[f"{_cls_def['counter']}_list"] = copy.deepcopy(schema)
+        _cls_def[f"{_cls_def['counter']}_list"]['referenced'] = True
+        schema = {'$ref': f"#/$defs/{_cls_def['counter']}_list"}
+        _cls_def['counter'] += 1
+
+    # Accept singletons
+    res = {'oneOf': [schema]}
     # Wrap type in list
     if f_type.is_list:
-        res = {
+        res['oneOf'].append({
             'type': 'array',
-            'items': res,
-        }
+            'items': schema,
+        })
     # Wrap type for optional
     if f_type.is_optional:
-        res = {
-            'oneOf': [
-                {
-                    'type': 'null'
-                },
-                res,
-            ]
-        }
+        res['oneOf'].append({'type': 'null'})
+
+    # Explicitly shortcut primitives
+    if len(f_type.types) == 1 and f_type.type in (str, bool, int, float):
+        key_name = f'{f_type.type}_{f_type.is_list}_{f_type.is_optional}'
```

Comment:
nit: make the key more human-readable

```suggestion
        key_name = f'{f_type.type.__qualname__}{'_list' if f_type.is_list else ''}{'_optiona' if f_type.is_optional else ''}'
```